File: site-src/examples/algorithm-deepening/pattern-lab-v08/structured_dp_trace.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

std::vector<int> knapsack(const std::vector<int>& weights, const std::vector<int>& values,
                          int capacity, bool descending) {
  if (weights.size() != values.size() || capacity < 0) throw std::invalid_argument("invalid knapsack input");
  for (const int weight : weights) if (weight <= 0) throw std::invalid_argument("invalid knapsack input");
  std::vector<int> dp(static_cast<std::size_t>(capacity + 1), 0);
  for (std::size_t item = 0; item < weights.size(); ++item) {
    if (descending) {
      for (int current = capacity; current >= weights[item]; --current) {
        dp[current] = std::max(dp[current], dp[current - weights[item]] + values[item]);
      }
    } else {
      for (int current = weights[item]; current <= capacity; ++current) {
        dp[current] = std::max(dp[current], dp[current - weights[item]] + values[item]);
      }
    }
  }
  return dp;
}

std::pair<long long, std::string> matrix_chain(const std::vector<int>& dimensions) {
  if (dimensions.size() < 2) throw std::invalid_argument("invalid matrix dimensions");
  for (const int dimension : dimensions) if (dimension <= 0) throw std::invalid_argument("invalid matrix dimensions");
  const std::size_t count = dimensions.size() - 1;
  std::vector<std::vector<long long>> cost(count, std::vector<long long>(count, 0));
  std::vector<std::vector<std::size_t>> split(count, std::vector<std::size_t>(count, 0));
  for (std::size_t length = 2; length <= count; ++length) {
    for (std::size_t left = 0; left + length <= count; ++left) {
      const std::size_t right = left + length - 1;
      long long best = std::numeric_limits<long long>::max();
      for (std::size_t middle = left; middle < right; ++middle) {
        const long long candidate = cost[left][middle] + cost[middle + 1][right]
          + static_cast<long long>(dimensions[left]) * dimensions[middle + 1] * dimensions[right + 1];
        if (candidate < best) {
          best = candidate;
          split[left][right] = middle;
        }
      }
      cost[left][right] = best;
    }
  }
  const auto build = [&](const auto& self, std::size_t left, std::size_t right) -> std::string {
    if (left == right) return "A" + std::to_string(left + 1);
    const std::size_t middle = split[left][right];
    return "(" + self(self, left, middle) + self(self, middle + 1, right) + ")";
  };
  return {cost[0][count - 1], build(build, 0, count - 1)};
}

void print_values(const std::vector<int>& values) {
  for (std::size_t index = 0; index < values.size(); ++index) {
    if (index > 0) std::cout << ',';
    std::cout << values[index];
  }
}

int main() {
  const auto table = knapsack({2,3,4,5}, {3,4,5,8}, 7, true);
  const int correct_single = knapsack({2}, {3}, 4, true).back();
  const int wrong_single = knapsack({2}, {3}, 4, false).back();
  const auto matrix = matrix_chain({10,30,5,60});
  std::cout << "knapsack_weights=2,3,4,5 values=3,4,5,8 capacity=7\n";
  std::cout << "knapsack_dp="; print_values(table); std::cout << '\n';
  std::cout << "knapsack_optimal=" << table.back() << '\n';
  std::cout << "forward_single_item=" << wrong_single << " correct_single_item=" << correct_single << '\n';
  std::cout << "matrix_dims=10,30,5,60\n";
  std::cout << "matrix_cost=" << matrix.first << " order=" << matrix.second << '\n';
  std::cout << "orders=capacity-descending,interval-length-ascending\n";
  std::cout << "invariants=item-used-at-most-once,subintervals-ready\n";
}

File: examples/algorithm-deepening/pattern-lab-v08/structured_dp_trace.py
from __future__ import annotations

from itertools import combinations
from math import inf
from typing import Sequence


def knapsack_01(weights: Sequence[int], values: Sequence[int], capacity: int) -> tuple[int, ...]:
    if len(weights) != len(values) or capacity < 0 or any(weight <= 0 for weight in weights):
        raise ValueError("invalid knapsack input")
    dp = [0] * (capacity + 1)
    for weight, value in zip(weights, values):
        for current in range(capacity, weight - 1, -1):
            dp[current] = max(dp[current], dp[current - weight] + value)
    return tuple(dp)


def knapsack_forward_wrong(weights: Sequence[int], values: Sequence[int], capacity: int) -> tuple[int, ...]:
    if len(weights) != len(values) or capacity < 0 or any(weight <= 0 for weight in weights):
        raise ValueError("invalid knapsack input")
    dp = [0] * (capacity + 1)
    for weight, value in zip(weights, values):
        for current in range(weight, capacity + 1):
            dp[current] = max(dp[current], dp[current - weight] + value)
    return tuple(dp)


def knapsack_bruteforce(weights: Sequence[int], values: Sequence[int], capacity: int) -> int:
    if len(weights) != len(values) or capacity < 0:
        raise ValueError("invalid knapsack input")
    best = 0
    for size in range(len(weights) + 1):
        for chosen in combinations(range(len(weights)), size):
            if sum(weights[index] for index in chosen) <= capacity:
                best = max(best, sum(values[index] for index in chosen))
    return best


def matrix_chain(dimensions: Sequence[int]) -> tuple[int, str]:
    if len(dimensions) < 2 or any(dimension <= 0 for dimension in dimensions):
        raise ValueError("invalid matrix dimensions")
    count = len(dimensions) - 1
    cost = [[0] * count for _ in range(count)]
    split = [[0] * count for _ in range(count)]
    for length in range(2, count + 1):
        for left in range(count - length + 1):
            right = left + length - 1
            best = inf
            for middle in range(left, right):
                candidate = (
                    cost[left][middle]
                    + cost[middle + 1][right]
                    + dimensions[left] * dimensions[middle + 1] * dimensions[right + 1]
                )
                if candidate < best:
                    best = candidate
                    split[left][right] = middle
            cost[left][right] = int(best)

    def build(left: int, right: int) -> str:
        if left == right:
            return f"A{left + 1}"
        middle = split[left][right]
        return f"({build(left, middle)}{build(middle + 1, right)})"

    return cost[0][count - 1], build(0, count - 1)


def fixed_report() -> str:
    weights, values, capacity = [2, 3, 4, 5], [3, 4, 5, 8], 7
    table = knapsack_01(weights, values, capacity)
    correct_single = knapsack_01([2], [3], 4)[-1]
    wrong_single = knapsack_forward_wrong([2], [3], 4)[-1]
    matrix_cost, matrix_order = matrix_chain([10, 30, 5, 60])
    return "\n".join(
        [
            "knapsack_weights=2,3,4,5 values=3,4,5,8 capacity=7",
            f"knapsack_dp={','.join(map(str, table))}",
            f"knapsack_optimal={table[-1]}",
            f"forward_single_item={wrong_single} correct_single_item={correct_single}",
            "matrix_dims=10,30,5,60",
            f"matrix_cost={matrix_cost} order={matrix_order}",
            "orders=capacity-descending,interval-length-ascending",
            "invariants=item-used-at-most-once,subintervals-ready",
        ]
    )


if __name__ == "__main__":
    print(fixed_report())

